src/amo/pages/AddonReviewList/index.js

```diff
@@ -63,7 +71,8 @@ type InternalProps = {|
   pageSize: number | null,
   reviewCount?: number,
   reviews?: Array<UserReviewType>,
-  areReviewsLoading: boolean,
+  siteUser: UserType | null,
+  siteUserCanReplyToReviews: ?boolean,
```

Comment:
Can you use `| void` instead? I like it better because it clearly shows that `undefined` as a value is important (which seems to be the case in `componentDidMount()`). `null` is [allowed here](https://flow.org/en/docs/types/maybe/), and it is not clear to me whether it is wanted.

That being said, do you use `undefined` here because the selector returns `undefined`? I don't really like this mix of values vs undefined vs null. I especially don't like `undefined` because if you forget a `return` or something, Flow won't complain.

We should really think about using a solid [Option/Maybe lib](https://github.com/alexanderjarvis/maybe)...

</ rant >

Reply:
Whoops, the question mark type was not intended. We could benefit from a lint rule to disallow question mark types in favor of explicit `null` and/or `void` types but there is a lot of existing code to fix.

> do you use undefined here because the selector returns undefined? I don't really like this mix of values vs undefined vs null.

I was allowing `undefined` because the reducer sets `canReplyToReviews` to `undefined` before it has been fetched it from the API. That is a pattern I'm used to: `undefined` usually means it has *not* been fetched from the API whereas `null` means it has been fetched from the API and does not exist. However, that isn't relevant here since the API result will always be a `boolean`. Thus, I changed this initialization to `null`.

>  I especially don't like undefined because if you forget a return or something, Flow won't complain.

If you unintentionally omit a return value from a function, Flow or lint usually complains. Lint will complain about inconsistent return values and Flow will complain if the returned value doesn't match another input type. But, yeah, JavaScript. I changed the initialization to `null` as suggested.

> We should really think about using a solid Option/Maybe lib...

I would rather stick to primitives, if we can. I think with Flow coverage it's pretty safe to make comparisons to primitives. We could clean up our consistency for how we do it, though.

Reply:
> That is a pattern I'm used to: `undefined` usually means it has _not_ been fetched from the API whereas `null` means it has been fetched from the API and does not exist.

The problem with this "pattern" is that it is your pattern, and no one is supposed to know it because I don't think it is documented or enforced. I know it because you left comments in the past, but contributors or other developers might not know it.

A lib for true Option/Maybe types would allow us to introduce this pattern (pretty much) in an robust and explicit manner.

Reply:
Yes, we definitely need to improve our docs for which patterns we use on the site. However, anyone who has tried to build an SSR compliant loading state on AMO should know the pattern by now.